### PR TITLE
added overflow checks and corresponding casts

### PR DESCRIPTION
**Pull request by makart19**
added mul overflow checks and casts were necessary for PCA, KMeans, KMeansInit

### CODE REVIEW

File: cpp/oneapi/dal/algo/kmeans/BUILD [hunk 1]
```diff
@@ -41,6 +52,7 @@ dal_test_suite(
 dal_test_suite(
     name = "tests",
     host_tests = [
+        ":common_tests",
         ":cpu_tests",
     ],
     dpc_tests = [
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/algo/kmeans/BUILD:58]: Why you do not include common tests to the `dpc_tests` if they are really common?

**Review by michael-smirnov — COMMENTED** [review]
Don't forget to add backport label

File: cpp/oneapi/dal/algo/kmeans/BUILD (hunk 1 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/algo/kmeans/BUILD:58]: It seems to me, common tests should be able to run on host only (w/o GPU)

Comment (makart19) [reply cpp/oneapi/dal/algo/kmeans/BUILD:58]: yes, from my perspective in common tests we are testing common parts/branches for CPU/GPU (like descriptor) on host, i.e. we can avoid launch on GPU.

File: cpp/oneapi/dal/algo/kmeans/common_kmeans_test.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#include "oneapi/dal/algo/kmeans/infer.hpp"
+#include "oneapi/dal/algo/kmeans/train.hpp"
+#include "oneapi/dal/table/row_accessor.hpp"
+
+using namespace oneapi;
+namespace kmn = oneapi::dal::kmeans;
+
+using kmn_desc = kmn::descriptor<float, kmn::method::lloyd_dense, kmn::task::clustering>;
+
+TEST(kmn_bad_arg_tests, set_cluster_count) {
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/kmeans/common_kmeans_test.cpp:27]: Maybe kmeans_bad_arg_tests?

File: cpp/oneapi/dal/algo/kmeans/BUILD (hunk 1 above)
Comment (michael-smirnov) [reply cpp/oneapi/dal/algo/kmeans/BUILD:58]: We have two interfaces of oneDAL: host and dpc++. When you say "common tests," I think these tests shall work for both interfaces - okay, if you will test in only on CPUs for dpc++ interface. But seems like you are testing the host interface only.

File: cpp/oneapi/dal/algo/kmeans/common_kmeans_test.cpp (hunk 2 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/kmeans/common_kmeans_test.cpp]: done

**Review by mchernov-intel — APPROVED** [review]

File: cpp/oneapi/dal/algo/kmeans/BUILD (hunk 1 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/kmeans/BUILD:58]: made tests renaming

**Pull request merged**